server/src/test/java/com/adobe/testing/s3mock/its/ListObjectV1MaxKeysIT.java

```
@@ -0,0 +1,97 @@
+    assertEquals(1, objectListing.getObjectSummaries().size());
+    assertEquals(1, objectListing.getMaxKeys());
+  }
+
+  @Test
+  public void returnsAllObjectsIfMaxKeysIsDefault() {
+    ListObjectsRequest request = new ListObjectsRequest().withBucketName(bucketName);
+    ObjectListing objectListing = s3Client.listObjects(request);
+
+    assertEquals(2, objectListing.getObjectSummaries().size());
+    assertEquals(1000, objectListing.getMaxKeys());
+  }
+
+  @Test
+  public void returnsAllObjectsIfMaxKeysEqualToAmountOfObjects() {
+    ListObjectsRequest request = new ListObjectsRequest().withBucketName(bucketName)
+            .withMaxKeys(2);
+    ObjectListing objectListing = s3Client.listObjects(request);
+
+    assertEquals(2, objectListing.getObjectSummaries().size());
+    assertEquals(2, objectListing.getMaxKeys());
+  }
+
+  @Test
+  public void returnsAllObjectsIfMaxKeysMoreThanAmountOfObjects() {
+    ListObjectsRequest request = new ListObjectsRequest().withBucketName(bucketName)
+            .withMaxKeys(3);
+    ObjectListing objectListing = s3Client.listObjects(request);
+
+    assertEquals(2, objectListing.getObjectSummaries().size());
+    assertEquals(3, objectListing.getMaxKeys());
+  }
+
+  @Test
+  public void returnsEmptyListIfMaxKeysIsZero() {
+    ListObjectsRequest request = new ListObjectsRequest().withBucketName(bucketName)
+            .withMaxKeys(0);
+    ObjectListing objectListing = s3Client.listObjects(request);
+
+    assertEquals(0, objectListing.getObjectSummaries().size());
+    assertEquals(0, objectListing.getMaxKeys());
+  }
+
+  @Test
+  public void returnsAllObjectsIfMaxKeysIsNegative() {
+    ListObjectsRequest request = new ListObjectsRequest().withBucketName(bucketName)
+            .withMaxKeys(-1);
+    ObjectListing objectListing = s3Client.listObjects(request);
+
+    // Apparently, the Amazon SDK rejects negative max keys, and by default it's 1000
+    assertEquals(2, objectListing.getObjectSummaries().size());
```

How does this test pass, if you throw the S3Exception on negative values in the implementation?

---

it looks like the Amazon SDK ignores negative `maxKeys` and doesn't send them to the server, so we get the default value of 1000. I throw the exception on the server side as a safety measure to fail fast.